src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java

```diff
@@ -1110,19 +1114,33 @@ public StructImpl downloadAndExtract(
     }
 
     StructImpl downloadResult = calculateDownloadResult(checksum, downloadedPath);
-    try {
-      if (downloadDirectory.exists()) {
-        downloadDirectory.deleteTree();
+    // Retry the deleteTree a while, if necessary, to cope with filesytems where unlinks are not atomic
+    Instant start = Instant.now();
+    Instant deadline = start.plus(Duration.ofSeconds(5));
```

Comment:
An easy objection is that 5 seconds is a long time, and I am not doing any sleeping or backoff in this loop, but my reasoning is: previously, this would be an immediate abort of the entire build, so the length of time or CPU usage of retrying something that was previously completely unhandled is a bit moot.  In a few tests in our environment, 90% of these circumstances are resolved in ~100ms but occasionally the file deletes are not reflected for over 500ms so I figured "better safe than sorry", and I didn't want to make this diff any more complicated than it already was.

Reply:
Could you give an example of such a filesystem? What exactly does it mean for unlink to not be atomic?

Reply:
Some distributed NFS appliances have this property, and they ack the delete before all the caches are updated. The symptom is that you rm a file, but it may not be absent in an immediate followup list.

Reply:
Another thought: Could we download to a temporary directory outside the repo that we can clean up asynchronously?

Reply:
That would work too, I was actually surprised the download was done into output_dir but I figured it was due to some hermeticity argument.

Reply:
To be clear, this is still rather uncommon - amongst hundreds/thousands of downloads, it happens a handful of times, but that’s plenty for aborting the entire build to be very painful.

Reply:
I guess that the download is happening into the repo dir only because of symmetry with separate `download` and `extract` invocations. 

@Wyverald What do you think?

Reply:
I could add similar logic to close() for the deleteTree there, but that one does not affect us and I wanted to make the most surgical PR I could.

Reply:
Is there anything I can do to move this along?  It solves a massive problem for us, and I am sure others who use similar filesystems.

Reply:
is `deleteTree` the only call that needs this special treatment?

either way, I think I'd prefer if you could extract this special handling into a separate method (`deleteTreeWithRetries` or something), and add some documentation explaining the rationale (preferably also linking to the GH issue). That way, it's less implementation burden for us.

Reply:
It's not really me using this function, it's the dependency downloading of `http_archive` and all the crate downloading of `rules_rust`.  But, if what you mean is that I should simply wrap up the retries into a function which is instead used by `downloadAndExtract` I am happy to do that.  I can do that right now one minute.

Reply:
For now I am adding it inline in the `StarlarkBaseExternalContext` since `Path.java` seems pretty closely aligned with the file system calls...

Reply:
thanks, this looks good to me.

> It's not really me using this function

I understand that -- it's just that having the extra logic _specific_ to this call of `deleteTree` invites the question of "why isn't this done elsewhere". Extracting the logic into a separate method makes it easier to apply the same retry logic to other call sites.